graph/iterator/lines.go

```diff
@@ -129,3 +132,87 @@ func (e *OrderedWeightedLines) WeightedLineSlice() []graph.WeightedLine {
 func (e *OrderedWeightedLines) Reset() {
 	e.idx = -1
 }
+
+// OrderedTemporalLines implements the graph.TemporalLines and graph.TemporalLineSlicer interfaces.
+// The iteration order of OrderedTemporalLines is the order of lines passed to
+// NewTemporalLineIterator.
+type OrderedTemporalLines struct {
+	idx   int
+	lines []graph.TemporalLine
+}
+
+// NewOrderedTemporalLines returns an OrderedTemporalLines initialized with the provided lines.
+func NewOrderedTemporalLines(lines []graph.TemporalLine) *OrderedTemporalLines {
+	return &OrderedTemporalLines{idx: -1, lines: lines}
+}
+
+// Len returns the remaining number of lines to be iterated over.
+func (e *OrderedTemporalLines) Len() int {
+	if e.idx >= len(e.lines) {
+		return 0
+	}
+	if e.idx <= 0 {
+		return len(e.lines)
+	}
+	return len(e.lines[e.idx:])
+}
+
+// Next returns whether the next call of TemporalLine will return a valid line.
+func (e *OrderedTemporalLines) Next() bool {
+	if uint(e.idx)+1 < uint(len(e.lines)) {
+		e.idx++
+		return true
+	}
+	e.idx = len(e.lines)
+	return false
+}
+
+// TemporalLine returns the current line of the iterator. Next must have been
+// called prior to a call to TemporalLine.
+func (e *OrderedTemporalLines) TemporalLine() graph.TemporalLine {
+	if e.idx >= len(e.lines) || e.idx < 0 {
+		return nil
+	}
+	return e.lines[e.idx]
+}
+
+// TemporalLineSlice returns all the remaining lines in the iterator and advances
+// the iterator.
+func (e *OrderedTemporalLines) TemporalLineSlice() []graph.TemporalLine {
+	if e.idx >= len(e.lines) {
+		return nil
+	}
+	idx := e.idx
+	if idx == -1 {
+		idx = 0
+	}
+	e.idx = len(e.lines)
+	return e.lines[idx:]
+}
+
+// Reset returns the iterator to its initial state.
+func (e *OrderedTemporalLines) Reset() {
+	e.idx = -1
+}
+
+func (e *OrderedTemporalLines) Swap (i, j int) {
+	e.lines[i], e.lines[j] = e.lines[j], e.lines[i]
+}
+
+func (e *OrderedTemporalLines) Less(i, j int) bool {
+	if e.lines[i].At() < e.lines[j].At() {
+		return true
+	}
+	if e.lines[i].Duration() < e.lines[j].Duration() {
+		return true
+	}
+	return false
+}
+
+func (e *OrderedTemporalLines) LineStream() graph.TemporalLines {
+	stream := make([]graph.TemporalLine, len(e.lines))
```

Comment:
I don't understand why this needs a copy. If you want to lazily sort, then just sort lazily on the unused edges. Otherwise just make it a requirement that the edges be sorted and do a `sort.IsSorted` when `New...` is called. If the latter is done then you can guarantee that the edges will always come out in order and this method can go away.